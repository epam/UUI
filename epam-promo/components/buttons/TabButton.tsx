import css from './TabButton.scss';
import styles from '../../assets/styles/colorvars/buttons/tabButton-colorvars.scss';
import { Button, ButtonProps } from '@epam/uui-components';
import { withMods } from '@epam/uui-core';
import { systemIcons } from '../../icons/icons';
import { getIconClass } from './helper';

export interface TabButtonMods {
    /** Tab button size */
    size?: '36' | '48' | '60';

    /** Show notification red dot */
    withNotify?: boolean;
}

function applyTabButtonMods(mods: TabButtonMods & ButtonProps) {
    return [css.root, css['size-' + (mods.size || '48')], mods.withNotify && css.uuiNotification, styles.tabButtonColor, ...getIconClass(mods)];
}

export const TabButton = withMods<ButtonProps, TabButtonMods>(Button, applyTabButtonMods, props => ({
    dropdownIcon: systemIcons['36'].foldingArrow,
    clearIcon: systemIcons['36'].clear,
    countPosition: 'right',
    ...props,
    rawProps: { role: 'tab', ...props.rawProps },
}));
