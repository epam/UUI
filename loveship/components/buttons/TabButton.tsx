import css from './TabButton.module.scss';
import * as uui from '@epam/uui';
import { withMods } from '@epam/uui-core';

export interface TabButtonMods extends uui.TabButtonProps {
    theme?: 'light' | 'dark';
}

function applyTabButtonMods(mods: TabButtonMods & uui.TabButtonProps) {
    return [mods.theme === 'dark' && css.themeDark];
}

export interface TabButtonProps extends uui.TabButtonProps, TabButtonMods {}

export const TabButton = withMods<uui.TabButtonProps, TabButtonMods>(uui.TabButton, applyTabButtonMods);
