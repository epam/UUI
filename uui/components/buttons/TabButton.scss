@use 'sass:math';
@use  '../../assets/styles/index' as *;

@mixin tab-button-color($color, $activeColor, $backgroundColor) {
    color: $color;
    fill: $color;
    background-color: $backgroundColor;

    &:hover, &:focus {
        color: $activeColor;
        fill: $activeColor;
    }

    &:global(.uui-active) {
        color: $activeColor;
        fill: $activeColor;
        border-color: $activeColor;
    }
}

$text-size-caption: 36px;
$text-size-count: 24px;

.root {
    @include link-button-horizontal-layout(36px);
    @include tab-button-color(var(--uui-tab-button-text), var(--uui-tab-button-text-active), transparent);
    cursor: pointer;
    padding-left: 12px;
    padding-right: 12px;
    position: relative;
    bottom: -1px;
    display: flex;
    align-items: center;
    font-family: var(--font-semibold);
    font-weight: normal;
    border-bottom: 3px solid transparent;

    &:global(.uui-active) {
        border-color: var(--uui-tab-button-text-active);
    }

    &:global(.uui-disabled) {
        @include tab-button-color(var(--uui-tab-button-text-disabled), var(--uui-tab-button-text-disabled), transparent);
    }

    &.uui-notification {
        &::after {
            content: '';
            margin-bottom: math.div($text-size-caption, 2);
            height: 6px;
            width: 6px;
            background: var(--uui-negative);
            box-shadow: 0 1px 8px 0 var(--uui-negative);
            border-radius: 50%;
        }
    }

    :global(.uui-count) {
        @include text-size($text-size-count, 0);
        letter-spacing: 0;
        color: var(--uui-tab-button-count);
        background-color: var(--uui-tab-button-count-bg);
        padding: 0 6px;
        margin-left: 6px;
        margin-right: 3px;
        border-radius: 9px;
    }

    :global(.uui-caption) {
        @include text-size($text-size-caption, 0);
    }

    // Sizes
    &.size-36 {
        height: 36px;
    }

    &.size-48 {
        height: 48px;
    }

    &.size-60 {
        height: 60px;
    }
}