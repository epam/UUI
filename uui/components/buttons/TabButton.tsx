import React from 'react';
import {
    cx,
    Icon,
    IDropdownToggler,
    IHasCaption,
    IHasIcon,
    uuiElement,
    uuiMarkers,
    Overwrite,
    ICanBeActive, useIsActive, uuiMod,
} from '@epam/uui-core';
import { Clickable, ClickableComponentProps, IconContainer } from '@epam/uui-components';
import { getIconClass } from './helper';
import { CountIndicator } from '../widgets/CountIndicator';
import { settings } from '../../settings';

import css from './TabButton.module.scss';

type TabButtonMods = {
    /**
     * Defines component size.
     * @default '48'
     */
    size?: '36' | '48' | '60';
    /** Defines is the component showing Notify */
    withNotify?: boolean;
};

export interface TabButtonModsOverride {}

/** Represents the properties of a TabButton component. */
export interface TabButtonProps extends Overwrite<TabButtonMods, TabButtonModsOverride>, ClickableComponentProps, IDropdownToggler, IHasIcon, IHasCaption, ICanBeActive {
    /** Call to clear toggler value */
    onClear?(e?: any): void;
    /** Icon for clear value button (usually cross) */
    clearIcon?: Icon;
    /** Icon for drop-down toggler */
    dropdownIcon?: Icon;
    /** Count value to be placed in component */
    count?: React.ReactNode;
}

export const TabButton = React.forwardRef<HTMLButtonElement | HTMLAnchorElement | HTMLSpanElement, TabButtonProps>((props, ref) => {
    const { isActive } = useIsActive({
        isLinkActive: props.isLinkActive,
        link: props.link,
        isActive: props.isActive,
    });

    const styles = [
        isActive && uuiMod.active,
        css.root,
        'uui-tab-button',
        `uui-size-${props.size || settings.tabButton.sizes.default}`,
        ...getIconClass(props),
        props.cx,
    ];

    const DropdownIcon = props.dropdownIcon ? props.dropdownIcon : settings.tabButton.icons.dropdownIcon;
    const ClearIcon = props.clearIcon ? props.clearIcon : settings.tabButton.icons.clearIcon;

    return (
        <Clickable
            { ...props }
            rawProps={ {
                role: 'tab',
                'aria-haspopup': props.isDropdown,
                'aria-expanded': props.isOpen,
                ...props.rawProps,
            } }
            cx={ styles }
            ref={ ref }
        >
            { props.icon && props.iconPosition !== 'right' && (
                <IconContainer
                    icon={ props.icon }
                    onClick={ !props.isDisabled ? props.onIconClick : undefined }
                />
            ) }
            { (props.caption || props.withNotify) && (
                <div className={ cx(uuiElement.caption) }>
                    { props.caption }
                    { props.withNotify && <div className={ css.withNotify } /> }
                </div>
            ) }
            { props.icon && props.iconPosition === 'right' && (
                <IconContainer icon={ props.icon } onClick={ !props.isDisabled ? props.onIconClick : undefined } />
            ) }
            { props.count !== undefined && props.count !== null && (
                <CountIndicator
                    color="neutral"
                    size={ settings.tabButton.sizes.countIndicatorMap[props.size || settings.tabButton.sizes.default] }
                    caption={ props.count }
                />
            ) }
            { props.isDropdown && (
                <IconContainer icon={ DropdownIcon } flipY={ props.isOpen } />
            )}
            { props.onClear && !props.isDisabled && (
                <IconContainer cx={ uuiMarkers.clickable } icon={ ClearIcon } onClick={ props.onClear } />
            ) }
        </Clickable>
    );
});
