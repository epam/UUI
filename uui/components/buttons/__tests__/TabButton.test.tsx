import React from 'react';
import { TabButton } from '../TabButton';
import { renderSnapshotWithContextAsync } from '@epam/uui-test-utils';
import { ReactComponent as calendarIcon } from '@epam/assets/icons/action-calendar-fill.svg';

describe('TabButton', () => {
    it('should be rendered correctly', async () => {
        const tree = await renderSnapshotWithContextAsync(<TabButton />);
        expect(tree).toMatchSnapshot();
    });

    it('should be rendered correctly with props', async () => {
        const tree = await renderSnapshotWithContextAsync(
            <TabButton
                onClick={ jest.fn }
                icon={ calendarIcon }
                isDisabled={ false }
                withNotify={ true }
                isActive={ true }
            />,
        );
        expect(tree).toMatchSnapshot();
    });
});
