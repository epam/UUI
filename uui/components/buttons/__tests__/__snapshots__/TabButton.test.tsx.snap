// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabButton should be rendered correctly 1`] = `
<DocumentFragment>
  <span
    class="uui-button-box uui-enabled root uui-tab-button uui-size-48 uui-no-left-icon uui-no-right-icon"
    role="tab"
    tabindex="-1"
  />
</DocumentFragment>
`;

exports[`TabButton should be rendered correctly with props 1`] = `
<DocumentFragment>
  <button
    aria-disabled="false"
    class="uui-button-box uui-enabled -clickable uui-active root uui-tab-button uui-size-48 uui-has-left-icon uui-no-right-icon"
    role="tab"
    tabindex="0"
    type="button"
  >
    <div
      class="container uui-icon uui-enabled"
    >
      <svg
        class=""
      />
    </div>
    <div
      class="uui-caption"
    >
      <div
        class="withNotify"
      />
    </div>
  </button>
</DocumentFragment>
`;
